Reorganize Architecture chapter

Andriciuc opened the pull request:
- add initial new topics, split from Architecture.md. Now the user can see the following topics easier: Overview, Usage guide, Encryption internals, Key provider management topics
- add introduction paragraph to some topics that were lacking it and improve clarity of the descriptions and content inside the topics
- update ## Key providers (principal key storage) topic as it is outdated
- fix links due to architecture.md being removed and updated TOC
- remove last mentions of WAL being beta (I know this is out of scope, apologies)

Andriciuc committed:
Reorganize Architecture chapter

- add initial new topics: overview, usage guide, encryption internals, key provider management topics
- move content for overview from architecture.md, pending others too

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into improve-architecture-chapter-docs

Andriciuc committed:
added information from architecture.md to each subtopic

Andriciuc committed:
modify intro to fit new structure

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into improve-architecture-chapter-docs

Andriciuc committed:
remove warning of WAL beta from wal_encryption variable

- improve overview with cards
- change title to match ToC and not feel disjointed
- remove architecture.md since it is split now
- fix links
- rewrite Key providers (principal key storage) topic in key-prov-management as it contains out of date info

nastena1606 commented on contrib/pg_tde/documentation/docs/architecture/usage-guide.md:
```diff
@@ -0,0 +1,165 @@
+# User interface
```
Why is this page under Architecture?

Andriciuc replied:
whoops, wrong title, fixed to # Usage guide

Also, you are correct! As I am splitting the old Arch chapter up, I haven't gotten the chance to move it to its new location under Technical reference -> Usage guide. 

I'll update it asap. 👍

nastena1606 commented on contrib/pg_tde/documentation/docs/advanced-topics/tech-reference.md:
```diff
@@ -1,7 +1,39 @@
-[Architecture](../architecture/architecture.md){.md-button} [GUC Variables](../variables.md){.md-button} [Functions](../functions.md){.md-button}
+<div data-grid markdown><div data-banner markdown>
+
+### :material-cog-refresh-outline: Architecture { .title }
```
You have Architecture on the index page and then here. With different content on a card. IMHO, it's too overcrowded with the same card, as if we insist the user MUST learn our architecture.

Andriciuc replied:
Yeah I was debating this really hard, should we remove this card and keep the rest or reinforce/go more in-depth about the content the user will learn about in this chapter?

So leaning towards removing it?

Andriciuc replied:
I have replaced the Architecture card with the recently added Usage guide, thank you very much for the feedback on this! It does feel better and not so overloaded towards seeing the Arch topic

Andriciuc committed:
update overview.md with updated customizable support and extension explanation

- add intro paragraph for Encryption architecture to reflect data at rest capabilities explained below it

Andriciuc committed:
move usage guide to Technical reference outside of Architecture

Andriciuc committed:
add intro paragraph for Usage guide and replace Architecture card with Usage guide info

Andriciuc committed:
update overview and internals with clearer text

- update variable for code branch

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into improve-architecture-chapter-docs

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into improve-architecture-chapter-docs

Andriciuc committed:
update text for clarity in SMGR API and WAL encryption

- add intro paragraph to key and key provider management

ImTheKai commented on contrib/pg_tde/documentation/docs/advanced-topics/usage-guide.md:
```diff
@@ -0,0 +1,169 @@
+
+* Add `pg_tde` to the `shared_preload_libraries` in `postgresql.conf` as this is required for the SMGR extensions
+* Execute `CREATE EXTENSION pg_tde` in the databases where they want to use encryption
+* Optionally, enable `pg_tde.wal_encrypt` in `postgresql.conf`
```
Why should I do that?

Andriciuc replied:
Updated to: Optionally, enable `pg_tde.wal_encrypt` in `postgresql.conf` to encrypt WAL writes

ImTheKai commented on contrib/pg_tde/documentation/docs/advanced-topics/usage-guide.md:
```diff
@@ -0,0 +1,169 @@
+* Add `pg_tde` to the `shared_preload_libraries` in `postgresql.conf` as this is required for the SMGR extensions
+* Execute `CREATE EXTENSION pg_tde` in the databases where they want to use encryption
+* Optionally, enable `pg_tde.wal_encrypt` in `postgresql.conf`
+* Optionally, disable `pg_tde.inherit_global_providers` in `postgresql.conf` (it is enabled by default)
```
Why should I do that?

Andriciuc replied:
Updated to: Optionally, disable `pg_tde.inherit_global_providers` in `postgresql.conf` if you do **not** want the database to use global key providers (it is enabled by default)

ImTheKai commented on contrib/pg_tde/documentation/docs/advanced-topics/usage-guide.md:
```diff
@@ -0,0 +1,169 @@
+
+## Create encrypted tables
+
+To create an encrypted table or modify an existing table to be encrypted, use the following commands:
```
I would split the example below into create and alter - if forsee (while there's not harm), that people will simply do both

Andriciuc replied:
True! Split this into two separate use cases:

## Create an encrypted table

To create an encrypted table, use the following command:

```sql
CREATE TABLE t1(a INT) USING tde_heap;
```

## Alter an encrypted table

To alter or modify an encrypted table, use the following command:

```sql
ALTER TABLE t1 SET ACCESS METHOD tde_heap;
```

ImTheKai commented on contrib/pg_tde/documentation/docs/architecture/encrypt-internals.md:
```diff
@@ -0,0 +1,75 @@
+`pg_tde` uses two kinds of keys for encryption:
+
+1. Internal keys to encrypt the data. They are stored in PostgreSQL's data directory under `$PGDATA/pg_tde`.
+2. Higher-level keys to encrypt internal keys. These keys are called *principal keys*. They are stored externally, in a Key Management System (KMS) using the key provider API.
```
I would directly call them principal keys instead of "higher-level"

Andriciuc replied:
Updated to: 

2. Principal keys, which encrypt internal keys, are stored externally in a Key Management System (KMS) using the key provider API.

Thank you!

ImTheKai commented on contrib/pg_tde/documentation/docs/architecture/key-prov-management.md:
```diff
@@ -0,0 +1,59 @@
+
+## Internal key regeneration
+
+Internal keys for tables, indexes and other data files are fixed once a file is created. There's no way to re-encrypt a file.
```
I would call it "are generated once a file is created" or another synonym, as "fixed" reads as something was broken

Andriciuc replied:
Good catch, fixed with the suggestion!

ImTheKai commented on contrib/pg_tde/documentation/docs/architecture/key-prov-management.md:
```diff
@@ -0,0 +1,59 @@
+
+There are workarounds for this, because operations that move the table data to a new file, such as `VACUUM FULL` or an `ALTER TABLE` that rewrites the file will create a new key for the new file, essentially rotating the internal key. This however means taking an exclusive lock on the table for the duration of the operation, which might not be desirable for huge tables.
+
+WAL internal keys are also fixed to the respective ranges. To generate a new WAL key you need to restart the database.
```
What does this mean? What is a range in this context?

Andriciuc replied:
I agree that range is incorrectly used in this context, I have rewritten as:

WAL internal keys are fixed once created. Every time the server (re)starts, a new WAL key is generated. If WAL encryption is enabled (using `pg_tde.wal_encrypt`), all WAL writes following the creation of the new key are encrypted with it until **another** key is generated at the next restart. This ensures that each WAL segment uses a consistent encryption key, without requiring you to manage key rotation manually.

ImTheKai commented on contrib/pg_tde/documentation/docs/architecture/key-prov-management.md:
```diff
@@ -0,0 +1,59 @@
+When you configure a key provider, `pg_tde`:
+
+* Uploads new principal keys when they are created
+* Retrieves principal keys from the provider when needed for decryption
```
When is it needed? Only on startup or restart of postgres, or how often?

Andriciuc replied:
Is the following correct as a solution?

* Retrieves principal keys from the provider when needed for decryption (for example, at startup, restart, or on cache miss)

ImTheKai replied:
I think that makes it clearer. The only question here would be, what on cache miss really means. How could that even happen? So if we know, add it, if we don't just drop it.

Andriciuc replied:
Dropped, and just left:

- Retrieves principal keys from the provider when needed for decryption (for example, at startup or restart)

Andriciuc committed:
Merge branch 'TDE_REL_17_STABLE' into improve-architecture-chapter-docs

Andriciuc committed:
updates based on feedback

Andriciuc committed:
Update key-prov-management.md
